### ISSUE TITLE
429 TooManyRequests errors while loading state

### ISSUE THREAD

**Issue by robin-wayve** [hashicorp/terraform-provider-azurerm#17747]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.2.4

### AzureRM Provider Version

3.13.0

### Affected Resource(s)/Data Source(s)

Any

### Terraform Configuration Files

We have about ~60 modules which I'm unable to share: however, I can say that I don't believe the number of resources we are managing is particularly large or unusual.

### Debug Output/Panic Output

```shell
Error: error loading state: Error retrieving keys for Storage Account "***": storage.AccountsClient#ListKeys: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="TooManyRequests" Message="The request is being throttled as the limit has been reached for operation type - Read_ObservationWindow_00:05:00. For more information, see - https://aka.ms/srpthrottlinglimits"
```

### Expected Behaviour

The throttling / rate limit headers should be respected as previously addressed in #202.

### Actual Behaviour

`terraform plan` and `terraform apply` infrequently fail due to 429 errors.

### Steps to Reproduce

Difficult to reproduce because I suspect it correlates with increased load (ie. when several engineers are working on our infrastructure in parallel running `plan`s and/or our CI is running an `apply`.

### Important Factoids

We use `terragrunt` to order how are modules are applied.

### References

#202

**Comment by magodo** [comment]
@robin-wayve Thank you for submitting this!

429 is handled by the underlying `go-autorest`, where since v14.0.0 it will not infinitely retry a request on 429, unless you opt in its old behavior:

> By default, the DoRetryForStatusCodes functions will no longer infinitely retry a request when the response returns an HTTP status code of 429 (StatusTooManyRequests). To opt in to the old behavior set autorest.Count429AsRetry to false.

https://github.com/Azure/go-autorest/blob/7dd32b67be4e6c9386b9ba7b1c44a51263f05270/CHANGELOG.md#breaking-changes

As you can see below, the provider has already opt in its old behavior (this has been set for a long while):

https://github.com/hashicorp/terraform-provider-azurerm/blob/546ba633b44b4271d2cb8bfb2e8572366d399c28/internal/clients/client.go#L228

Meanwhile, I've verified in my local that I have basically following config:

```hcl
data "azurerm_storage_account" "example" {
  count               = 200
  name                = "myaccount"
  resource_group_name = "rg1"
}
```

Then I run `terraform plan -parallelism=200` (multiple times), and observed when 429 occurs, the provider correctly wait for the indicated interval (by `retry-after`) and retry and succeed.

**Comment by robin-wayve** [comment]
I'm trying to reproduce with `TF_LOG=DEBUG` on, I'll let you know if/when I succeed.

I should have mentioned that I added `-refresh=false` to our runs recently in an effort to reduce requests to Azure but I still saw this problem afterwards.

**Comment by tombuildsstuff** [comment]
cc @manicminer ^

**Comment by manicminer** [comment]
@robin-wayve Thanks for the debug log. That confirms as suspected, that autorest is only retrying on 429 up to 3 times, and you're hitting a rate limit on your storage account.

Unfortunately, the only way to ameliorate this would be to increase the retry count, or maybe even set `autorest.Count429AsRetry = false` - however, such a fix would have to go into Terraform Core and would only be usable in the next Terraform release (at the soonest) - by which point we may be further along in our wider efforts to migrate to our own SDK.

In the meantime I would suggest looking at other ways to work around the rate limiting on your storage account, i.e.

- reducing the number of parallel runs being bottlenecked on the same storage account (Terragrunt likely a contributing factor here)
- distribute your states across more storage accounts
- ask Azure support to increase your API limits

**Comment by robin-wayve** [comment]
> ask Azure support to increase your API limits

I tried this one, and got a hard "no". 😂

**Comment by manicminer** [comment]
Let's keep this open, we'll track this issue and its resolution whether that takes the form of the autorest workaround or migrating to our own SDK :)

**Comment by robin-wayve** [comment]
Another workaround we haven't tried yet, but I wanted to note: move state to a postgres database.

**Comment by robin-wayve** [comment]
I think there is another fairly simple solution to this which did not occur to me until @Tom-Newton pointed something out.

This problem happened in our case because we are using a Service Principal, and it seems like the backend's default behaviour is to fetch an access key for the storage account (on every `init`, `plan` and `apply` per module): this is the thing that gets rate limited when you do it too often.

However, you can specify `ARM_ACCESS_KEY`, and when this is present, the backend doesn't make the `ListKeys` requests. This didn't work out of the box for all of our Terraform modules, but I can see a path to working around the issues I did see (mostly to do with `remote_state`). It's made things significantly quicker when we don't end up waiting for long `Retry-After` periods.

**Comment by robin-wayve** [comment]
Can confirm once I worked through the blockers to using `ARM_ACCESS_KEY` this problem appears to have disappeared for us.

**Comment by Hyper200** [comment]
Is this still an issue for anyone? With the release configuration in azure you cannot use the standard terraform jobs in devops while trying to use ARM_ACCESS_KEY as it expects to be given a service principal or identity to be used to grab the storage account keys. It also makes key rotation very difficult.

**Comment by katbyte** [comment]
Thank you for taking the time to report this issue, and apologies that it has sat without a resolution for so long.

This bug was reported against an early major version of the AzureRM provider. The provider has since moved on to v5.x, and the affected functionality has often been significantly reworked across those major releases, so bug reports from legacy major versions prior to v4.0 are generally no longer actionable as written.

As part of a cleanup of historical issues we are closing bug reports that target legacy provider versions. If you are still running into this behaviour on the current major version, please open a new issue with up-to-date reproduction details (feel free to reference this one) so it can be triaged with current context.

Thank you again for the report, and for your patience.
